core/src/main/java/org/apache/brooklyn/core/sensor/DurationSinceSensor.java

```diff
@@ -0,0 +1,108 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.brooklyn.core.sensor;
+
+import java.util.concurrent.Callable;
+
+import org.apache.brooklyn.api.entity.Entity;
+import org.apache.brooklyn.api.entity.EntityLocal;
+import org.apache.brooklyn.api.sensor.AttributeSensor;
+import org.apache.brooklyn.config.ConfigKey;
+import org.apache.brooklyn.core.config.ConfigKeys;
+import org.apache.brooklyn.core.effector.AddSensor;
+import org.apache.brooklyn.feed.function.FunctionFeed;
+import org.apache.brooklyn.feed.function.FunctionPollConfig;
+import org.apache.brooklyn.util.core.config.ConfigBag;
+import org.apache.brooklyn.util.time.Duration;
+
+import com.google.common.base.Supplier;
+import com.google.common.reflect.TypeToken;
+
+public class DurationSinceSensor extends AddSensor<Duration> {
+
+    private static final Supplier<Long> CURRENT_TIME_SUPPLIER = new CurrentTimeSupplier();
+
+    public static final ConfigKey<Supplier<Long>> EPOCH_SUPPLIER = ConfigKeys.builder(new TypeToken<Supplier<Long>>() {})
+            .name("duration.since.epochsupplier")
+            .description("The source of time from which durations are measured. Defaults to System.currentTimeMillis.")
+            .defaultValue(CURRENT_TIME_SUPPLIER)
+            .build();
+
+    public static final ConfigKey<Supplier<Long>> TIME_SUPPLIER = ConfigKeys.builder(new TypeToken<Supplier<Long>>() {})
+            .name("duration.since.timesupplier")
+            .description("The source of the current time. Defaults to System.currentTimeMillis.")
+            .defaultValue(CURRENT_TIME_SUPPLIER)
+            .build();
+
+    private final Supplier<Long> epochSupplier;
+    private final Supplier<Long> timeSupplier;
+    private final AttributeSensor<Long> epochSensor;
+
+    public DurationSinceSensor(ConfigBag params) {
+        super(params);
+        epochSupplier = params.get(EPOCH_SUPPLIER);
+        timeSupplier = params.get(TIME_SUPPLIER);
+        epochSensor = Sensors.newLongSensor(sensor.getName() + ".epoch");
+    }
+
+    @Override
+    public void apply(final EntityLocal entity) {
+        super.apply(entity);
+
+        if (entity.sensors().get(epochSensor) == null) {
+            entity.sensors().set(epochSensor, epochSupplier.get());
```

Comment:
I think if epochSupplier returns null it should default to using CURRENT_TIME_SUPPLIER. It's surely a programming error. I've updated the description of the config key to indicate this.